presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java

```diff
@@ -511,12 +510,10 @@ public HiveRecordWriter(
             // todo adapt input types to the file types as Hive does
             for (int fileIndex = 0; fileIndex < fileColumnNames.size(); fileIndex++) {
                 String columnName = fileColumnNames.get(fileIndex);
-                Type fileColumnType = fileColumnTypes.get(fileIndex);
-                int inputIndex = inputColumnNames.indexOf(columnName);
-                Type inputType = inputColumnTypes.get(inputIndex);
+                HiveType fileColumnHiveType = fileColumnHiveTypes.get(fileIndex);
+                HiveType inputHiveType = inputColumnMap.get(columnName).getHiveType();
 
-                // HACK: solve coercion in generic way
-                if (!typesMatchForInsert(fileColumnType, inputType)) {
+                if (!fileColumnHiveType.equals(inputHiveType)) {
```

Comment:
Short answer: because prest-hive tests were failing :)

Because `STRING` type was mapped to `VARCHAR` without parameters, thus `fileColumnType` was evaluated to `varchar(MAX_INT)`. That meant, you wouldn't be able to insert `varchar(x)` for `x < MAX_INT`, because types were not equal.

With your change we should be fine for now, until we introduce `varchar(x)` support to presto-hive. Correct me if I am  wrong, with your change Presto type `varchar(x)` for any `x` will be mapped to `string` hive type and that's why this equality check works? Once we add support for `varchar(x)` in presto-hive the same problem with equality comparisons will come back so I'm not sure whether we should leave the old check or to replace it with equality comparison for now.

Btw, if you decide to drop `typesMatchForInsert` you probably can delete that method, because as I remember it was only used here.